File: frontend/lib/objects-of-issue/defaults.ts
/**
 * Empty-record factories for Objects of the Issue (Increment O1).
 *
 * Every money / count / percentage field starts as `''` (never `0`, never `null`) and every
 * repeatable record receives a stable `crypto.randomUUID()` id so React keys and cross-section
 * links survive re-renders and round-trips.
 */

import type {
  AcquisitionsSubsidiariesJvsAndInvestments,
  BorrowingRepaymentItem,
  CapexItem,
  CapitalExpenditureFacilitiesAndExpansion,
  DeploymentScheduleRow,
  ExpensesGcpMonitoringAndConfirmations,
  InvestmentItem,
  IssueExpenseItem,
  IssueObject,
  MeansOfFinanceAndDeploymentSchedule,
  MeansOfFinanceRow,
  ObjectsOfIssueConfirmations,
  ObjectsOfIssuePayload,
  ObjectsRegisterAndAllocation,
  ProceedsAndFundingSummary,
  WorkingCapitalAndBorrowingRepayment,
} from '@/lib/schemas/objects-of-issue';
import { OBJECTS_OF_ISSUE_SCHEMA_VERSION } from '@/lib/schemas/objects-of-issue';

function newId(id?: string): string {
  return id ?? crypto.randomUUID();
}

/* -------------------------------------------------------------------------- */
/* 1. Proceeds & funding summary                                              */
/* -------------------------------------------------------------------------- */

export function createEmptyProceedsAndFundingSummary(): ProceedsAndFundingSummary {
  return {
    declaredOfferType: '',
    freshIssueGrossProceeds: '',
    estimatedIssueRelatedExpenses: '',
    issueMadeToRaiseFundsForObjects: '',
    schemeOfArrangementInvolved: '',
    offerForSaleProceedsNote: '',
    notes: '',
  };
}

/* -------------------------------------------------------------------------- */
/* 2. Objects register & allocation                                          */
/* -------------------------------------------------------------------------- */

export function createEmptyIssueObject(id?: string): IssueObject {
  return {
    id: newId(id),
    objectName: '',
    objectCategory: '',
    description: '',
    estimatedCost: '',
    amountFromNetProceeds: '',
    amountFromInternalAccruals: '',
    amountFromOtherSources: '',
    appraisalStatus: '',
    appraisingAgencyName: '',
    expectedUtilisationPeriod: '',
    priorityRank: '',
    notes: '',
  };
}

export function createEmptyObjectsRegisterAndAllocation(): ObjectsRegisterAndAllocation {
  return {
    objects: [],
    objectsAreFinalised: '',
    notes: '',
  };
}

/* -------------------------------------------------------------------------- */
/* 3. Capital expenditure, facilities & expansion                            */
/* -------------------------------------------------------------------------- */

export function createEmptyCapexItem(id?: string): CapexItem {
  return {
    id: newId(id),
    itemType: '',
    description: '',
    location: '',
    relatedObjectId: '',
    estimatedCost: '',
    expectedCommissioningDate: '',
    quotationSource: '',
    relatedPartyPurchase: '',
    governmentApprovalsRequired: '',
    approvalsStatus: '',
    notes: '',
  };
}

export function createEmptyCapitalExpenditureFacilitiesAndExpansion(): CapitalExpenditureFacilitiesAndExpansion {
  return {
    capexItems: [],
    notApplicableNote: '',
    notes: '',
  };
}

/* -------------------------------------------------------------------------- */
/* 4. Working capital & borrowing repayment                                  */
/* -------------------------------------------------------------------------- */

export function createEmptyBorrowingRepaymentItem(id?: string): BorrowingRepaymentItem {
  return {
    id: newId(id),
    lenderName: '',
    loanType: '',
    outstandingAmount: '',
    amountProposedForRepayment: '',
    interestRatePercentage: '',
    isRelatedPartyLender: '',
    repaymentRationale: '',
    notes: '',
  };
}

export function createEmptyWorkingCapitalAndBorrowingRepayment(): WorkingCapitalAndBorrowingRepayment {
  return {
    workingCapitalRequirementAmount: '',
    workingCapitalMethodology: '',
    workingCapitalAppraisalStatus: '',
    borrowingRepaymentItems: [],
    notes: '',
  };
}

/* -------------------------------------------------------------------------- */
/* 5. Acquisitions, subsidiaries, JVs & investments                          */
/* -------------------------------------------------------------------------- */

export function createEmptyInvestmentItem(id?: string): InvestmentItem {
  return {
    id: newId(id),
    targetEntityName: '',
    transactionType: '',
    relatedObjectId: '',
    estimatedAmount: '',
    proposedStakePercentage: '',
    definitiveAgreementStatus: '',
    regulatoryApprovalsRequired: '',
    regulatoryApprovalDetails: '',
    isRelatedPartyTransaction: '',
    rationale: '',
    notes: '',
  };
}

export function createEmptyAcquisitionsSubsidiariesJvsAndInvestments(): AcquisitionsSubsidiariesJvsAndInvestments {
  return {
    investmentItems: [],
    notes: '',
  };
}

/* -------------------------------------------------------------------------- */
/* 6. Means of finance & deployment schedule                                 */
/* -------------------------------------------------------------------------- */

export function createEmptyMeansOfFinanceRow(id?: string): MeansOfFinanceRow {
  return {
    id: newId(id),
    source: '',
    amount: '',
    notes: '',
  };
}

export function createEmptyDeploymentScheduleRow(id?: string): DeploymentScheduleRow {
  return {
    id: newId(id),
    periodLabel: '',
    amountToBeDeployed: '',
    notes: '',
  };
}

export function createEmptyMeansOfFinanceAndDeploymentSchedule(): MeansOfFinanceAndDeploymentSchedule {
  return {
    meansOfFinanceRows: [],
    deploymentScheduleRows: [],
    fundingTieUpStatus: '',
    fundingTieUpDetails: '',
    notes: '',
  };
}

/* -------------------------------------------------------------------------- */
/* 7. Expenses, GCP, monitoring & confirmations                              */
/* -------------------------------------------------------------------------- */

export function createEmptyIssueExpenseItem(id?: string): IssueExpenseItem {
  return {
    id: newId(id),
    expenseCategory: '',
    estimatedAmount: '',
    notes: '',
  };
}

export function createEmptyObjectsOfIssueConfirmations(): ObjectsOfIssueConfirmations {
  return {
    objectsServeBonafideBusinessPurposes: false,
    noPartOfProceedsBenefitsRelatedPartiesBeyondDisclosed: false,
    deploymentScheduleIsManagementEstimate: false,
    shortfallToBeMetFromInternalAccrualsOrOtherSources: false,
    meansOfFinanceExcludingIssueProceedsAlreadyTiedUp: false,
    monitoringAndUtilisationCertificationRequirementUnderstood: false,
    professionalReviewRemainsRequired: false,
  };
}

/** Back-compat alias. */
export const createEmptyObjectConfirmations = createEmptyObjectsOfIssueConfirmations;

export function createEmptyExpensesGcpMonitoringAndConfirmations(): ExpensesGcpMonitoringAndConfirmations {
  return {
    issueExpenseItems: [],
    generalCorporatePurposesAmount: '',
    monitoringAgencyRequired: '',
    monitoringAgencyName: '',
    monitoringAgencyStatus: '',
    confirmations: createEmptyObjectsOfIssueConfirmations(),
    notes: '',
  };
}

/* -------------------------------------------------------------------------- */
/* Payload                                                                     */
/* -------------------------------------------------------------------------- */

export function createEmptyObjectsOfIssuePayload(): ObjectsOfIssuePayload {
  return {
    schemaVersion: OBJECTS_OF_ISSUE_SCHEMA_VERSION,
    proceedsAndFundingSummary: createEmptyProceedsAndFundingSummary(),
    objectsRegisterAndAllocation: createEmptyObjectsRegisterAndAllocation(),
    capitalExpenditureFacilitiesAndExpansion:
      createEmptyCapitalExpenditureFacilitiesAndExpansion(),
    workingCapitalAndBorrowingRepayment: createEmptyWorkingCapitalAndBorrowingRepayment(),
    acquisitionsSubsidiariesJvsAndInvestments:
      createEmptyAcquisitionsSubsidiariesJvsAndInvestments(),
    meansOfFinanceAndDeploymentSchedule: createEmptyMeansOfFinanceAndDeploymentSchedule(),
    expensesGcpMonitoringAndConfirmations: createEmptyExpensesGcpMonitoringAndConfirmations(),
  };
}

File: backend/app/modules/objects_issue/defaults.py
"""Empty canonical Objects of the Issue payload (schemaVersion 1) — mirrors frontend O1 exactly."""

from __future__ import annotations

from copy import deepcopy
from typing import Any
from uuid import uuid4

from app.modules.objects_issue.constants import SCHEMA_VERSION


def _new_id(id_: str | None = None) -> str:
    return id_ or str(uuid4())


def create_empty_proceeds_and_funding_summary() -> dict[str, Any]:
    return {
        "declaredOfferType": "",
        "freshIssueGrossProceeds": "",
        "estimatedIssueRelatedExpenses": "",
        "issueMadeToRaiseFundsForObjects": "",
        "schemeOfArrangementInvolved": "",
        "offerForSaleProceedsNote": "",
        "notes": "",
    }


def create_empty_issue_object(id_: str | None = None) -> dict[str, Any]:
    return {
        "id": _new_id(id_),
        "objectName": "",
        "objectCategory": "",
        "description": "",
        "estimatedCost": "",
        "amountFromNetProceeds": "",
        "amountFromInternalAccruals": "",
        "amountFromOtherSources": "",
        "appraisalStatus": "",
        "appraisingAgencyName": "",
        "expectedUtilisationPeriod": "",
        "priorityRank": "",
        "notes": "",
    }


def create_empty_objects_register_and_allocation() -> dict[str, Any]:
    return {
        "objects": [],
        "objectsAreFinalised": "",
        "notes": "",
    }


def create_empty_capex_item(id_: str | None = None) -> dict[str, Any]:
    return {
        "id": _new_id(id_),
        "itemType": "",
        "description": "",
        "location": "",
        "relatedObjectId": "",
        "estimatedCost": "",
        "expectedCommissioningDate": "",
        "quotationSource": "",
        "relatedPartyPurchase": "",
        "governmentApprovalsRequired": "",
        "approvalsStatus": "",
        "notes": "",
    }


def create_empty_capital_expenditure_facilities_and_expansion() -> dict[str, Any]:
    return {
        "capexItems": [],
        "notApplicableNote": "",
        "notes": "",
    }


def create_empty_borrowing_repayment_item(id_: str | None = None) -> dict[str, Any]:
    return {
        "id": _new_id(id_),
        "lenderName": "",
        "loanType": "",
        "outstandingAmount": "",
        "amountProposedForRepayment": "",
        "interestRatePercentage": "",
        "isRelatedPartyLender": "",
        "repaymentRationale": "",
        "notes": "",
    }


def create_empty_working_capital_and_borrowing_repayment() -> dict[str, Any]:
    return {
        "workingCapitalRequirementAmount": "",
        "workingCapitalMethodology": "",
        "workingCapitalAppraisalStatus": "",
        "borrowingRepaymentItems": [],
        "notes": "",
    }


def create_empty_investment_item(id_: str | None = None) -> dict[str, Any]:
    return {
        "id": _new_id(id_),
        "targetEntityName": "",
        "transactionType": "",
        "relatedObjectId": "",
        "estimatedAmount": "",
        "proposedStakePercentage": "",
        "definitiveAgreementStatus": "",
        "regulatoryApprovalsRequired": "",
        "regulatoryApprovalDetails": "",
        "isRelatedPartyTransaction": "",
        "rationale": "",
        "notes": "",
    }


def create_empty_acquisitions_subsidiaries_jvs_and_investments() -> dict[str, Any]:
    return {
        "investmentItems": [],
        "notes": "",
    }


def create_empty_means_of_finance_row(id_: str | None = None) -> dict[str, Any]:
    return {
        "id": _new_id(id_),
        "source": "",
        "amount": "",
        "notes": "",
    }


def create_empty_deployment_schedule_row(id_: str | None = None) -> dict[str, Any]:
    return {
        "id": _new_id(id_),
        "periodLabel": "",
        "amountToBeDeployed": "",
        "notes": "",
    }


def create_empty_means_of_finance_and_deployment_schedule() -> dict[str, Any]:
    return {
        "meansOfFinanceRows": [],
        "deploymentScheduleRows": [],
        "fundingTieUpStatus": "",
        "fundingTieUpDetails": "",
        "notes": "",
    }


def create_empty_issue_expense_item(id_: str | None = None) -> dict[str, Any]:
    return {
        "id": _new_id(id_),
        "expenseCategory": "",
        "estimatedAmount": "",
        "notes": "",
    }


def create_empty_objects_of_issue_confirmations() -> dict[str, Any]:
    return {
        "objectsServeBonafideBusinessPurposes": False,
        "noPartOfProceedsBenefitsRelatedPartiesBeyondDisclosed": False,
        "deploymentScheduleIsManagementEstimate": False,
        "shortfallToBeMetFromInternalAccrualsOrOtherSources": False,
        "meansOfFinanceExcludingIssueProceedsAlreadyTiedUp": False,
        "monitoringAndUtilisationCertificationRequirementUnderstood": False,
        "professionalReviewRemainsRequired": False,
    }


def create_empty_expenses_gcp_monitoring_and_confirmations() -> dict[str, Any]:
    return {
        "issueExpenseItems": [],
        "generalCorporatePurposesAmount": "",
        "monitoringAgencyRequired": "",
        "monitoringAgencyName": "",
        "monitoringAgencyStatus": "",
        "confirmations": create_empty_objects_of_issue_confirmations(),
        "notes": "",
    }


def empty_payload() -> dict[str, Any]:
    return {
        "schemaVersion": SCHEMA_VERSION,
        "proceedsAndFundingSummary": create_empty_proceeds_and_funding_summary(),
        "objectsRegisterAndAllocation": create_empty_objects_register_and_allocation(),
        "capitalExpenditureFacilitiesAndExpansion": (
            create_empty_capital_expenditure_facilities_and_expansion()
        ),
        "workingCapitalAndBorrowingRepayment": create_empty_working_capital_and_borrowing_repayment(),
        "acquisitionsSubsidiariesJvsAndInvestments": (
            create_empty_acquisitions_subsidiaries_jvs_and_investments()
        ),
        "meansOfFinanceAndDeploymentSchedule": create_empty_means_of_finance_and_deployment_schedule(),
        "expensesGcpMonitoringAndConfirmations": (
            create_empty_expenses_gcp_monitoring_and_confirmations()
        ),
    }


def clone_empty_payload() -> dict[str, Any]:
    return deepcopy(empty_payload())
